File: internal/api/schedule.go
package api

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/christopherime/schedularr/internal/api/gen"
	"github.com/christopherime/schedularr/internal/external/tunarr"
	"github.com/christopherime/schedularr/internal/scheduler"
	"github.com/christopherime/schedularr/internal/service"
)

// Schedule day-window bounds shared by POST /generate, POST /apply (both
// take GenerateRequest, whose days field the OpenAPI schema marks default
// 7, minimum 1, maximum 30) and GET /schedule?days=N (same bounds, in its
// query parameter). As with GetHistory's days handling
// (internal/api/history.go), oapi-codegen's generated bindings do not
// enforce the schema's default/minimum/maximum, so both are applied here.
const (
	defaultScheduleDays = 7
	minScheduleDays     = 1
	maxScheduleDays     = 30
)

// GenerateSchedule implements gen.ServerInterface. It always runs a dry
// run (service.Options.Apply: false) regardless of the request body --
// only POST /apply mutates anything; this route exists purely to preview
// a plan.
func (h *Handlers) GenerateSchedule(w http.ResponseWriter, r *http.Request) {
	body, ok := h.decodeGenerateRequest(w, r)
	if !ok {
		return
	}
	h.runSchedule(w, r, scheduleRequest{days: body.Days, channelID: body.ChannelId, apply: false, op: "generate_schedule"})
}

// ApplySchedule implements gen.ServerInterface. Unlike GenerateSchedule, it
// runs with service.Options.Apply: true, so a successful response means
// Runner.Run pushed the plan to Tunarr (UpdateSchedule per channel) and
// committed the engine's pending state.
func (h *Handlers) ApplySchedule(w http.ResponseWriter, r *http.Request) {
	body, ok := h.decodeGenerateRequest(w, r)
	if !ok {
		return
	}
	h.runSchedule(w, r, scheduleRequest{days: body.Days, channelID: body.ChannelId, apply: true, op: "apply_schedule"})
}

// GetSchedule implements gen.ServerInterface. It always dry-runs (there is
// no request body on a GET, and no apply concept for this route). channel_id
// scopes planning exactly like GenerateRequest.channel_id does on the POST
// routes -- the web guide's SCOPE control rides on this parameter.
func (h *Handlers) GetSchedule(w http.ResponseWriter, r *http.Request, params gen.GetScheduleParams) {
	h.runSchedule(w, r, scheduleRequest{days: params.Days, channelID: params.ChannelId, op: "get_schedule"})
}

// decodeGenerateRequest decodes a GenerateRequest body for GenerateSchedule
// and ApplySchedule. Unlike CreateBlock/UpdateBlock's BlockWrite (whose
// OpenAPI requestBody is required and whose fields are required too), the
// OpenAPI schema does not mark GenerateRequest's body required and every
// one of its fields is optional -- so a request with no body at all (the
// decode failing with io.EOF) is valid input, not a 400: it just means
// "use every default."
func (h *Handlers) decodeGenerateRequest(w http.ResponseWriter, r *http.Request) (gen.GenerateRequest, bool) {
	var body gen.GenerateRequest
	dec := json.NewDecoder(r.Body)
	dec.DisallowUnknownFields()
	if err := dec.Decode(&body); err != nil {
		if errors.Is(err, io.EOF) {
			return body, true
		}
		WriteProblem(w, r, http.StatusBadRequest, "invalid request body", err.Error())
		return gen.GenerateRequest{}, false
	}
	return body, true
}

// scheduleRequest bundles runSchedule's inputs: GenerateSchedule and
// ApplySchedule build one from a decoded JSON body, GetSchedule from its
// query parameters. op identifies the calling handler for
// writeScheduleRunnerError's server-side log line (e.g.
// "generate_schedule").
type scheduleRequest struct {
	days      *int
	channelID *string
	apply     bool
	op        string
}

// runSchedule validates days, resolves the optional channelID, calls
// Deps.Sched.Run, and writes the resulting plan -- the shared body behind
// all three schedule endpoints, which differ only in where their
// parameters come from (query string vs. JSON body) and whether apply is
// set.
func (h *Handlers) runSchedule(w http.ResponseWriter, r *http.Request, req scheduleRequest) {
	days := defaultScheduleDays
	if req.days != nil {
		days = *req.days
	}
	if days < minScheduleDays || days > maxScheduleDays {
		WriteProblem(w, r, http.StatusBadRequest, "invalid days parameter",
			fmt.Sprintf("days must be between %d and %d, got %d", minScheduleDays, maxScheduleDays, days))
		return
	}

	channelID := ""
	if req.channelID != nil {
		channelID = *req.channelID
	}

	result, err := h.d.Sched.Run(r.Context(), service.Options{Days: days, ChannelID: channelID, Apply: req.apply})
	if err != nil {
		h.writeScheduleRunnerError(w, r, req.op, err)
		return
	}

	writeJSON(w, http.StatusOK, planResultToGen(result))
}

// writeScheduleRunnerError logs err server-side and writes a 502
// problem+json response with a short, fixed detail. Unlike ListChannels'
// Tunarr-connectivity errors (safe to echo verbatim -- see tunarr.go), a
// Runner.Run failure can wrap a raw store/driver error (e.g.
// Engine.Commit's error path during apply), so -- matching
// logAndWriteInternalError's information-leak convention -- the real error
// only ever reaches the server log, never the response body. op identifies
// which handler code path failed (e.g. "generate_schedule").
func (h *Handlers) writeScheduleRunnerError(w http.ResponseWriter, r *http.Request, op string, err error) {
	h.d.Logger.Error("schedule generation failed",
		"op", op,
		"request_id", RequestIDFromContext(r.Context()),
		"error", err,
	)
	WriteProblem(w, r, http.StatusBadGateway, "schedule generation failed", "unable to generate or apply the schedule")
}

// planResultToGen converts a service.Result (the domain representation) to
// a gen.PlanResult (the API wire shape). Warnings is omitted entirely
// (nil, not an empty array) when result.Warnings is empty, matching
// OpenAPI's `warnings` being optional -- a caller with nothing to report
// sees no key at all, not a distracting `"warnings": []` on every
// response.
func planResultToGen(result *service.Result) gen.PlanResult {
	channels := make(map[string][]gen.ScheduledSlot, len(result.Channels))
	for channelID, slots := range result.Channels {
		list := make([]gen.ScheduledSlot, 0, len(slots))
		for _, slot := range slots {
			list = append(list, slotToGen(slot))
		}
		channels[channelID] = list
	}

	plan := gen.PlanResult{Applied: result.Applied, Channels: channels}
	if len(result.Warnings) > 0 {
		warnings := make([]gen.Warning, 0, len(result.Warnings))
		for _, w := range result.Warnings {
			warnings = append(warnings, warningToGen(w))
		}
		plan.Warnings = &warnings
	}
	return plan
}

// warningToGen converts a scheduler.Warning into a gen.Warning.
func warningToGen(w scheduler.Warning) gen.Warning {
	blockName := w.BlockName
	occurrenceStart := w.OccurrenceStart
	blockingBlockName := w.BlockingBlockName
	return gen.Warning{
		BlockName:         &blockName,
		OccurrenceStart:   &occurrenceStart,
		BlockingBlockName: &blockingBlockName,
	}
}

// slotToGen converts a scheduler.ScheduledSlot into a gen.ScheduledSlot.
// Block reuses specToGen (blocks.go), the same scheduler.Block ->
// gen.BlockSpec conversion ListBlocks/GetBlock/... use. Programs is the
// typed ScheduledProgram projection (a hard swap from the old
// additionalProperties passthrough, per the no-legacy policy): each
// program's own start_time is computed here as the slot's start plus the
// cumulative durations of everything before it in the lineup -- the same
// wall-clock math Tunarr applies when it plays the lineup back to back.
func slotToGen(slot scheduler.ScheduledSlot) gen.ScheduledSlot {
	start := slot.StartTime
	end := slot.EndTime
	block := specToGen(slot.Block)

	programs := make([]gen.ScheduledProgram, 0, len(slot.Programs))
	cursor := slot.StartTime
	for _, p := range slot.Programs {
		programs = append(programs, programToGen(p, cursor))
		cursor = cursor.Add(time.Duration(p.GetDurationMs()) * time.Millisecond)
	}

	return gen.ScheduledSlot{
		StartTime: &start,
		EndTime:   &end,
		Block:     &block,
		Programs:  &programs,
	}
}

// programToGen projects one tunarr.Program onto the wire's typed
// ScheduledProgram at its computed air time. Optional fields are omitted
// (nil) rather than sent as zero values: a flex placeholder has no type
// worth naming beyond what Tunarr set, and season/episode 0 means "not an
// episode", not S00E00.
func programToGen(p tunarr.Program, startTime time.Time) gen.ScheduledProgram {
	out := gen.ScheduledProgram{
		Title:      p.Title,
		DurationMs: p.GetDurationMs(),
		StartTime:  startTime,
	}
	if p.Type != "" {
		t := p.Type
		out.Type = &t
	}
	if p.SeasonNumber > 0 {
		s := p.SeasonNumber
		out.Season = &s
	}
	if p.EpisodeNumber > 0 {
		e := p.EpisodeNumber
		out.Episode = &e
	}
	return out
}

File: web/assets/ts/pages/schedule.ts
// Schedule page ("/schedule/"): preview a dry-run plan (POST /api/v1/generate)
// per channel, then push it to Tunarr (POST /api/v1/apply) behind an explicit
// confirmation dialog -- the web equivalent of the CLI's `--yes` gate
// (cmd/generate.go's checkApplyGate). Bundled separately via "page_js", same
// pattern as dashboard.ts/blocks.ts.
//
// Contract notes that shape the code below:
//
//   1. `/generate` always dry-runs (PlanResult.applied is always false) and
//      `/apply` always mutates (applied: true on success) -- both take the
//      identical GenerateRequest body (days, optional channel_id). This page
//      never invents a third request shape: buildRequestBody() is shared by
//      both preview() and confirmApply().
//   2. `/apply` does NOT replay the client's previewed plan -- it independently
//      re-runs Runner.Run server-side (internal/service/schedule.go) with the
//      same parameters. "Apply with the same body as the preview" (the
//      binding contract's wording) means the same *request*, not a cached
//      plan payload; this file follows that literally.
//   3. `channel_id`, when set, restricts which blocks get planned at all
//      (README's "Schedule Endpoints" notes) -- an empty/omitted value means
//      every channel. This page always omits `channel_id` from the wire body
//      when the control resolves to "" ("All channels"), never sending an
//      empty string, matching blocks.ts's "omit rather than send a blank
//      placeholder" convention.
//   4. ScheduledSlot.programs is the typed ScheduledProgram shape
//      (api/openapi.yaml, v0.5.1 hard swap): title/duration_ms/start_time
//      required, type/season/episode optional. `title` is contractually a
//      string; the render helper below still normalizes a blank one to an
//      em dash so the expandable list never shows an empty row.
import { apiGet, apiPath, apiSend, onReauth } from "../runtime/api.ts";
import type { ApiRequestJSON, ApiResponse } from "../runtime/api.ts";
import { channelHint as channelHintText, channelLabel, channelPlate, loadChannels } from "../runtime/channels.ts";
import type { Channel, PlateParts } from "../runtime/channels.ts";
import { toProblemView } from "../runtime/errors.ts";
import type { ProblemView } from "../runtime/errors.ts";
import { formatLocal, plural } from "../runtime/format.ts";
import { initShell } from "../runtime/shell.ts";
import type { components } from "../gen/types";

initShell();

type PlanResult = ApiResponse<"generateSchedule", 200>;
type ScheduledSlot = components["schemas"]["ScheduledSlot"];
type Warning = components["schemas"]["Warning"];
type GenerateBody = ApiRequestJSON<"generateSchedule">;

declare const Alpine: {
  data<T extends object>(name: string, factory: () => T): void;
};

// Same double-init defense as dashboard.ts/blocks.ts: Alpine.data()'s init()
// is auto-invoked, so nothing on this page also wires x-init="init()" to it.
let started = false;

// Problem rendering note: this page's binding contract calls for the
// API's own title *and* detail rendered as two distinct pieces -- the 502
// case specifically names "schedule generation failed" (the server's
// literal Problem.title, internal/api/schedule.go:writeScheduleRunnerError)
// as something the UI must show verbatim. The shared ProblemView/ui-problem
// idiom does exactly that (title line + detail line + REF), so this page
// now uses the runtime's toProblemView like every other page.

// ---- controls -> wire request -------------------------------------------

interface Controls {
  days: string; // free text, like blocks.ts's numeric fields -- parsed once at request time, never via x-model.number
  channelId: string; // "" = All channels; select value or free-text fallback value
}

/** The exact {days, channelId} a preview or apply request was built from.
 * canApply() compares this against the *live* controls (via requestSignature())
 * to decide whether the current plan still matches what's on screen -- the
 * binding contract's "changing days/channel invalidates" rule. */
interface RequestSignature {
  days: number;
  channelId: string;
}

/** Clamps free-text days input to the API's documented [1, 30] range
 * (api/openapi.yaml GenerateRequest.days minimum/maximum), defaulting to 7
 * (the schema's own default) for blank/non-finite input. This is the
 * "client-side clamp" half of the binding contract; a value that somehow
 * still slips out of range is caught by the API's own 400 (the "rely on API
 * 400" half) -- toProblem()/previewError render that like any other
 * problem+json response. */
export function clampDays(raw: string): number {
  const trimmed = raw.trim();
  // Blank input must take the schema default (7), not fall through
  // Number("") === 0 into the 1-day clamp -- the latter contradicted both
  // this function's contract and the field's own "defaults to 7" hint
  // (caught by web/tests/schedule.test.ts during the v0.5.0 refactor).
  if (trimmed === "") return 7;
  const n = Number(trimmed);
  if (!Number.isFinite(n)) return 7;
  return Math.min(30, Math.max(1, Math.round(n)));
}

/** Builds the GenerateRequest body preview() and confirmApply() both send --
 * one shape, shared, per the contract note above. channel_id is omitted
 * (never sent as "") when the scope is "All channels". */
function buildRequestBody(sig: RequestSignature): GenerateBody {
  const body: GenerateBody = { days: sig.days };
  if (sig.channelId !== "") body.channel_id = sig.channelId;
  return body;
}

/** Orders channel sections the way their plates read: by resolved channel
 * number first (a plan section headed `CH 04 · HORROR` sorting on raw
 * UUID looks arbitrary), then name, then raw id as the final tiebreak.
 * Channels the cache can't resolve (or that carry no number) sort after
 * numbered ones, by name/id -- exported for direct testing, same
 * convention as clampDays above. */
export function channelOrder(aId: string, bId: string, channels: Channel[]): number {
  const a = channels.find((c) => c.id === aId);
  const b = channels.find((c) => c.id === bId);
  const aNum = a?.number ?? Number.POSITIVE_INFINITY;
  const bNum = b?.number ?? Number.POSITIVE_INFINITY;
  if (aNum !== bNum) return aNum - bNum;
  const byName = (a?.name ?? "").localeCompare(b?.name ?? "");
  if (byName !== 0) return byName;
  return aId.localeCompare(bId);
}

// ---- program rendering ---------------------------------------------------

type ScheduledProgram = components["schemas"]["ScheduledProgram"];

/** Renders a program's title (see contract note 4 above): a blank title
 * renders as an em dash rather than an empty row, so the expandable
 * list's item count always matches programCount(). */
function programTitle(p: ScheduledProgram): string {
  return p.title.trim() !== "" ? p.title : "—";
}

interface ChannelEntry {
  id: string;
  slots: ScheduledSlot[];
}

interface ScheduleState {
  controls: Controls;

  channelsLoading: boolean;
  channelsError: string | null;
  channels: Channel[];

  previewing: boolean;
  previewError: ProblemView | null;
  plan: PlanResult | null;
  previewedRequest: RequestSignature | null;

  applying: boolean;
  applyError: ProblemView | null;
  appliedAt: string | null;

  init(): void;
  loadChannels(): Promise<void>;
  channelLabel(c: Channel): string;
  channelHint(): string;
  plate(id: string): PlateParts;

  requestSignature(): RequestSignature;
  canApply(): boolean;

  preview(): Promise<void>;
  requestApply(): void;
  cancelApply(force?: boolean): void;
  confirmApply(): Promise<void>;

  channelEntries(): ChannelEntry[];
  programCount(slot: ScheduledSlot): number;
  programTitles(slot: ScheduledSlot): string[];
  slotsLabel(entry: ChannelEntry): string;
  programsLabel(slot: ScheduledSlot): string;
  formatLocal(iso: string | undefined): string;

  warnings(): Warning[];
  warningsHeading(): string;
  warningLabel(w: Warning): string;

  applyScopeLabel(): string;
  applyConfirmBody(): string;
  appliedSummary(): string;
}

// $refs is an Alpine magic (https://alpinejs.dev/magics/refs), injected at
// runtime, not part of the object literal Alpine.data()'s factory returns --
// same ThisType erasure trick blocks.ts uses for $nextTick.
interface WithRefs {
  $refs: { confirmDialog: HTMLDialogElement };
}

document.addEventListener("alpine:init", () => {
  Alpine.data(
    "schedule",
    (): ScheduleState & ThisType<ScheduleState & WithRefs> => ({
      controls: { days: "7", channelId: "" },

      channelsLoading: true,
      channelsError: null,
      channels: [],

      previewing: false,
      previewError: null,
      plan: null,
      previewedRequest: null,

      applying: false,
      applyError: null,
      appliedAt: null,

      init() {
        if (started) return;
        started = true;
        void this.loadChannels();
        // Arming a new token re-fires the channel load; a failed preview
        // is an action, not a load, so it stays behind its own button.
        onReauth(() => {
          if (this.channelsError) void this.loadChannels();
        });
      },

      async loadChannels() {
        this.channelsLoading = true;
        this.channelsError = null;
        try {
          this.channels = await loadChannels();
        } catch (err) {
          const problem = toProblemView(err);
          this.channelsError = problem.detail ?? problem.title;
          this.channels = [];
        } finally {
          this.channelsLoading = false;
        }
      },

      channelLabel,

      plate(id) {
        return channelPlate(id, this.channels);
      },

      // Shared hint logic (runtime/channels.ts's channelHint); this page's
      // manual-entry wording additionally notes that blank still means
      // "All channels" -- the one real difference from the blocks editor's
      // channel field.
      channelHint() {
        return channelHintText(
          this.channelsLoading,
          this.channelsError,
          this.channels,
          "enter a channel ID manually, or leave blank for all channels",
        );
      },

      requestSignature() {
        return { days: clampDays(this.controls.days), channelId: this.controls.channelId.trim() };
      },

      // Apply is armed only when: nothing is in flight, a plan exists, and
      // that plan was generated for exactly the controls currently on
      // screen. Editing days/channel after a preview changes
      // requestSignature()'s result without touching previewedRequest, so
      // this flips false immediately -- Alpine re-evaluates it reactively
      // since both are read here.
      canApply() {
        if (this.previewing || this.applying) return false;
        if (!this.plan || !this.previewedRequest) return false;
        const current = this.requestSignature();
        return this.previewedRequest.days === current.days && this.previewedRequest.channelId === current.channelId;
      },

      async preview() {
        const sig = this.requestSignature();
        this.previewing = true;
        this.previewError = null;
        this.applyError = null;
        try {
          const result = await apiSend<PlanResult>(
            "POST",
            apiPath("/generate"),
            buildRequestBody(sig),
          );
          this.plan = result;
          this.previewedRequest = sig;
          this.appliedAt = null;
        } catch (err) {
          // Clearing the plan (rather than leaving a stale one on screen
          // under the error) matters here specifically: a plan generated
          // for a *different* scope sitting next to a fresh error could
          // read as still valid, even though canApply() would already
          // block it.
          this.previewError = toProblemView(err);
          this.plan = null;
          this.previewedRequest = null;
        } finally {
          this.previewing = false;
        }
      },

      requestApply() {
        if (!this.canApply()) return;
        this.applyError = null;
        this.$refs.confirmDialog.showModal();
      },

      // State-level guard, not just the view layer's :disabled/backdrop
      // checks in list.html: refuses to close the dialog while an apply is
      // in flight, same "explicit re-entrancy guard" convention blocks.ts
      // uses for pendingId (toggleEnabled/performDelete). This is the last
      // line of defense against a dismissed-but-still-mutating dialog --
      // :disabled on Cancel/close-X and the backdrop handler's `!applying`
      // check exist too, but neither is trustworthy on its own (a disabled
      // attribute is a browser-enforced UI convention, not a guarantee this
      // method itself can rely on). confirmApply() passes force:true for
      // its own two closes, both of which happen deliberately while
      // `applying` is still true (it flips false in the `finally` clause
      // that runs after).
      cancelApply(force = false) {
        if (this.applying && !force) return;
        this.$refs.confirmDialog.close();
      },

      // Sends the SAME body preview() sent (buildRequestBody over the
      // signature the shown plan was generated from), never a body built
      // from live controls -- if the operator could somehow reach this
      // method with stale controls, the applied scope must still match
      // what the confirmation dialog told them, not whatever is in the
      // fields right now.
      async confirmApply() {
        const sig = this.previewedRequest;
        if (!sig) {
          this.cancelApply();
          return;
        }
        this.applying = true;
        try {
          const result = await apiSend<PlanResult>(
            "POST",
            apiPath("/apply"),
            buildRequestBody(sig),
          );
          this.plan = result;
          this.appliedAt = new Date().toISOString();
          // Force a fresh preview before the next apply, even if the
          // operator immediately re-opens the dialog against unchanged
          // controls -- an applied plan is a past action, not a standing
          // offer to repeat it with one more click.
          this.previewedRequest = null;
          this.cancelApply(true);
        } catch (err) {
          this.applyError = toProblemView(err);
          this.cancelApply(true);
        } finally {
          this.applying = false;
        }
      },

      // Server order is per-block, not per-channel: the engine concatenates
      // each block's own occurrences (internal/scheduler/engine.go's
      // GenerateForTimeRange iterates blocks, appending each one's slots to
      // its channel's list), so a channel fed by two blocks (e.g. a morning
      // block and an evening block) arrives grouped by block, not sorted by
      // time. The binding contract calls for "slots chronological", so each
      // channel's slots are sorted here -- start_time is an RFC 3339
      // date-time (api/openapi.yaml's ScheduledSlot.start_time), and RFC
      // 3339's fixed field widths make its string form sort identically to
      // its chronological order, so a plain lexicographic compare is exact,
      // not an approximation.
      channelEntries() {
        if (!this.plan) return [];
        return Object.entries(this.plan.channels)
          .map(([id, slots]) => ({
            id,
            slots: [...slots].sort((a, b) => (a.start_time ?? "").localeCompare(b.start_time ?? "")),
          }))
          .sort((a, b) => channelOrder(a.id, b.id, this.channels));
      },

      programCount(slot) {
        return (slot.programs ?? []).length;
      },

      programTitles(slot) {
        return (slot.programs ?? []).map(programTitle);
      },

      slotsLabel(entry) {
        return plural(entry.slots.length, "slot");
      },

      programsLabel(slot) {
        return plural(this.programCount(slot), "program");
      },

      formatLocal,

      // Exact wording the binding contract requires: "Apply ALL channels"
      // vs "Apply channel <id>" -- callers concatenate "Apply " + this.
      // Falls back to the live signature if somehow called before
      // previewedRequest is set (defensive; requestApply() never opens the
      // dialog without it).
      applyScopeLabel() {
        const id = (this.previewedRequest ?? this.requestSignature()).channelId;
        return id === "" ? "ALL channels" : `channel ${id}`;
      },

      // The confirmation dialog's body sentence. Exact required substring
      // for the empty-plan case ("nothing to apply for this scope") per the
      // binding contract; otherwise a real count drawn from the shown plan,
      // never a guess.
      applyConfirmBody() {
        if (!this.plan) return "";
        const channelCount = Object.keys(this.plan.channels).length;
        if (channelCount === 0) return "There is nothing to apply for this scope.";
        const slotCount = Object.values(this.plan.channels).reduce((sum, slots) => sum + slots.length, 0);
        // "Applies", matching the dialog's own title ("Apply " +
        // applyScopeLabel()), the Confirm Apply button, and
        // appliedSummary()'s "Applied ..." result below -- see the copy
        // audit's Apply/Applied terminology decision.
        return `This applies ${plural(slotCount, "slot")} across ${plural(channelCount, "channel")} to Tunarr.`;
      },

      // Every dropped occurrence from the last preview/apply (both
      // /generate and /apply return PlanResult.warnings the same way) --
      // conflict resolution on the server (internal/scheduler/engine.go's
      // resolveConflicts) picked a higher- (or equal-, first-come-)
      // priority occurrence on the same channel instead. Previously this
      // was only visible in a server-side INFO log line.
      warnings() {
        return this.plan?.warnings ?? [];
      },

      warningsHeading() {
        return `${plural(this.warnings().length, "occurrence")} not scheduled — conflict with a higher-priority block`;
      },

      warningLabel(w) {
        const block = w.block_name ?? "A block";
        const blocker = w.blocking_block_name ?? "a higher-priority block";
        const when = this.formatLocal(w.occurrence_start);
        return `${block} lost to ${blocker} at ${when} and was not scheduled.`;
      },

      appliedSummary() {
        if (!this.plan || !this.plan.applied) return "";
        const channelCount = Object.keys(this.plan.channels).length;
        const when = this.appliedAt ? this.formatLocal(this.appliedAt) : "";
        if (channelCount === 0) {
          return when ? `Applied — nothing to apply for this scope (${when}).` : "Applied — nothing to apply for this scope.";
        }
        const slotCount = Object.values(this.plan.channels).reduce((sum, slots) => sum + slots.length, 0);
        const summary = `Applied ${plural(slotCount, "slot")} across ${plural(channelCount, "channel")}`;
        return when ? `${summary} at ${when}.` : `${summary}.`;
      },
    }),
  );
});
